### PR TITLE
Implement Kumaraswamy Distribution

### PR DESCRIPTION
**Pull request by vishwakftw**
This PR implements the Kumaraswamy distribution.

cc: @fritzo @alicanb @sdaulton

### CODE REVIEW

File: torch/distributions/kumaraswamy.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        new.concentration0 = self.concentration0.expand(batch_shape)
+        return super(Kumaraswamy, self).expand(batch_shape, _instance=new)
+
+    @property
+    def mean(self):
+        return _moments(self.concentration1, self.concentration0, 1)
+
+    @property
+    def variance(self):
+        return _moments(self.concentration1, self.concentration0, 2) - torch.pow(self.mean, 2)
```
Comment (sdaulton) [inline torch/distributions/kumaraswamy.py:61]: Why not compute and use log moments and use logsumexp these calculations?

File: torch/distributions/kumaraswamy.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    arg_constraints = {'concentration1': constraints.positive, 'concentration0': constraints.positive}
+    support = constraints.unit_interval
+    has_rsample = True
+
+    def __init__(self, concentration1, concentration0, validate_args=None):
+        self.concentration1, self.concentration0 = broadcast_all(concentration1, concentration0)
+        finfo = torch.finfo(self.concentration0.dtype)
+        base_dist = Uniform(torch.full_like(self.concentration0, 0),
+                            torch.full_like(self.concentration0, 1))
+        transforms = [AffineTransform(loc=1., scale=-torch.ones_like(self.concentration1)),
```
Comment (sdaulton) [inline torch/distributions/kumaraswamy.py:43]: why allocated a tensor here, rather than just pass -1?

File: torch/distributions/utils.py [hunk 3]
```diff
@@ -5,6 +5,10 @@
 from typing import Dict, Any
 
 
+euler_constant = 0.57721566490153286060  # Euler Mascheroni Constant
```
Comment (sdaulton) [inline torch/distributions/utils.py:8]: Why not read this from torch.distributions.gumbel?

File: torch/distributions/kumaraswamy.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        return _moments(self.concentration1, self.concentration0, 1)
+
+    @property
+    def variance(self):
+        return _moments(self.concentration1, self.concentration0, 2) - torch.pow(self.mean, 2)
+
+    def entropy(self):
+        t1 = (1 - self.concentration1.reciprocal())
+        t0 = (1 - self.concentration0.reciprocal())
+        H1 = torch.digamma(self.concentration0 + 1) + euler_constant
```
Comment (sdaulton) [inline torch/distributions/kumaraswamy.py:66]: Perhaps call this H0 for readability?

**Comment by sdaulton** [comment]
Thanks for putting this up!

File: torch/distributions/utils.py (hunk 3 above)
Comment (vishwakftw) [reply torch/distributions/utils.py:8]: I think it's good SWE practice to isolate constants into a separate file?

File: torch/distributions/kumaraswamy.py (hunk 4 above)
Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py:66]: Oh yes, sorry about that.

File: torch/distributions/kumaraswamy.py (hunk 1 above)
Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py:61]: I don't see an immediate benefit, but let me try benchmarking.

File: torch/distributions/utils.py (hunk 3 above)
Comment (sdaulton) [reply torch/distributions/utils.py:8]: Ah nice I missed that you updated the gumbel file to import this one

File: torch/distributions/kumaraswamy.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    arg_constraints = {'concentration1': constraints.positive, 'concentration0': constraints.positive}
+    support = constraints.unit_interval
+    has_rsample = True
+
+    def __init__(self, concentration1, concentration0, validate_args=None):
+        self.concentration1, self.concentration0 = broadcast_all(concentration1, concentration0)
+        finfo = torch.finfo(self.concentration0.dtype)
+        base_dist = Uniform(torch.full_like(self.concentration0, 0),
+                            torch.full_like(self.concentration0, 1))
+        transforms = [AffineTransform(loc=1., scale=-1.),
```
Comment (fritzo) [inline torch/distributions/kumaraswamy.py:43]: Why is the first AffineTransform needed? Aren't you simply flipping `Uniform(0,1)`?

**Review by fritzo — COMMENTED** [review]
Tests look good to me. I haven't verified the math.

File: torch/distributions/kumaraswamy.py (hunk 5 above)
Comment (vishwakftw) [reply torch/distributions/kumaraswamy.py:43]: Ah yes. I didn't notice 2 years ago, thanks.

File: torch/distributions/utils.py [hunk 6]
```diff
@@ -5,6 +5,10 @@
...
+
+
```
Comment (fritzo) [inline torch/distributions/utils.py:10]: I think flake8 will complain about three empty lines; you can just remove one.

**Review by fritzo — COMMENTED** [review]
LGTM after fixing lint error.

**Comment by vishwakftw** [comment]
Oh I'm so sorry for missing the lint comment @ezyang @fritzo . Thanks for importing.

**Pull request closed without merging**